File: src/main/java/io/ahenteti/blog/service/converter/UserConverter.java
package io.ahenteti.blog.service.converter;

import io.ahenteti.blog.exception.InvalidApplicationStateException;
import io.ahenteti.blog.model.api.user.AuthorApiResponse;
import io.ahenteti.blog.model.api.user.CurrentUserApiResponse;
import io.ahenteti.blog.model.api.user.GetUsersPageApiRequest;
import io.ahenteti.blog.model.api.user.UserApiResponse;
import io.ahenteti.blog.model.api.user.UsersPageApiResponse;
import io.ahenteti.blog.model.api.user.ValidGetUsersPageApiRequest;
import io.ahenteti.blog.model.core.user.EUserRole;
import io.ahenteti.blog.model.core.user.User;
import io.ahenteti.blog.model.core.user.UsersPage;
import io.ahenteti.blog.model.core.user.oauth2.IOAuth2User;
import io.ahenteti.blog.model.entity.RoleEntity;
import io.ahenteti.blog.model.entity.UserEntity;
import io.ahenteti.blog.service.dao.repository.RoleRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.ArrayList;
import java.util.function.Supplier;
import java.util.stream.Collectors;

@Service
public class UserConverter {

    @Autowired
    private RoleRepository roleRepository;

    public User toUser(UserEntity entity) {
        User res = new User();
        res.setId(entity.getId());
        res.setUsername(entity.getUsername());
        res.setAvatarUrl(entity.getAvatar());
        res.setProvider(entity.getProvider());
        res.setJoinAt(entity.getJoinAt());
        entity.getRoles().forEach(role -> res.getRoles().add(EUserRole.from(role.getName())));
        return res;
    }

    public User toUser(IOAuth2User user) {
        User res = new User();
        res.setId(user.getPrimaryKey());
        res.setUsername(user.getUsername());
        res.setAvatarUrl(user.getAvatarUrl());
        res.setProvider(user.getProvider());
        res.setRoles(user.getRoles());
        return res;
    }

    public UserEntity toUserEntity(IOAuth2User user) {
        UserEntity res = new UserEntity();
        res.setId(null);
        res.setUsername(user.getUsername());
        res.setAvatar(user.getAvatarUrl());
        res.setProvider(user.getProvider());
        RoleEntity role = roleRepository.findByName("USER").orElseThrow(throwInvalidApplicationStateException());
        res.getRoles().add(role);
        res.setJoinAt(Instant.now());
        return res;
    }

    public CurrentUserApiResponse toCurrentUserApiResponse(IOAuth2User user) {
        CurrentUserApiResponse res = new CurrentUserApiResponse();
        res.setUsername(user.getUsername());
        res.setAvatarUrl(user.getAvatarUrl());
        user.getRoles().forEach(role -> res.getRoles().add(role.getValue()));
        return res;
    }

    public AuthorApiResponse toAuthorApiResponse(User user) {
        AuthorApiResponse res = new AuthorApiResponse();
        res.setUsername(user.getUsername());
        res.setAvatarUrl(user.getAvatarUrl());
        return res;
    }

    public UserApiResponse toUserApiResponse(User user) {
        UserApiResponse res = new UserApiResponse();
        res.setUsername(user.getUsername());
        res.setAvatarUrl(user.getAvatarUrl());
        res.setProvider(user.getProvider());
        res.setJoinAtIso8601(user.getJoinAt().toString());
        return res;
    }

    private Supplier<InvalidApplicationStateException> throwInvalidApplicationStateException() {
        return () -> new InvalidApplicationStateException("USER role not found in database");
    }

    public GetUsersPageApiRequest toGetUsersPageApiRequest(IOAuth2User user, Integer page, Integer size, String sortBy, String sortDirection, String filter) {
        GetUsersPageApiRequest res = new GetUsersPageApiRequest();
        res.setUser(user);
        res.setPage(page);
        res.setSize(size);
        res.setSortBy(sortBy);
        res.setSortDirection(sortDirection);
        res.setFilter(filter);
        return res;

    }

    public UsersPageApiResponse toUsersPageApiResponse(UsersPage usersPage) {
        UsersPageApiResponse res = new UsersPageApiResponse();
        res.setPage(usersPage.getPage());
        res.setSize(usersPage.getSize());
        res.setTotalItems(usersPage.getTotalItems());
        res.setItems(usersPage.getItems().stream().map(this::toUserApiResponse).collect(Collectors.toList()));
        return res;
    }

    public UsersPage toUsersPage(Page<UserEntity> users, ValidGetUsersPageApiRequest request) {
        UsersPage res = new UsersPage();
        res.setPage(request.getPage());
        res.setSize(request.getSize());
        res.setSortBy(request.getSortByValue());
        res.setTotalItems(users.getTotalElements());
        res.setItems(users.getContent().stream().map(this::toUser).collect(Collectors.toCollection(ArrayList::new)));
        return res;
    }
}

File: src/main/frontend/src/app/modules/user/services/user.converter.ts
import { Injectable } from "@angular/core";
import {
  AuthorApiResponse,
  CurrentUserApiResponse,
  UserApiResponse,
  GetUsersPageApiRequest,
  UsersPageApiResponse,
} from "../models/user.external.models";
import {
  Author,
  CurrentUser,
  User,
  UsersPage,
} from "../models/user.internal.models";

@Injectable()
export class UserConverter {
  toAuthor(author: AuthorApiResponse): Author {
    return {
      username: author.username,
      avatarUrl: author.avatarUrl,
    };
  }

  toCurrentUser(user: CurrentUserApiResponse): CurrentUser {
    return {
      username: user.username,
      avatarUrl: user.avatarUrl,
      isAuthenticated: user.username !== undefined && user.username !== null,
      isAdmin: user.roles.includes("ADMIN"),
    };
  }

  toUser(user: UserApiResponse): User {
    return {
      username: user.username,
      avatarUrl: user.avatarUrl,
      provider: user.provider,
      joinAt: new Date(user.joinAtIso8601),
    };
  }

  toGetUsersPageApiRequest(
    filter: string,
    page: number,
    size = 5,
    sortBy = "username",
    sortDirection = "asc"
  ): GetUsersPageApiRequest {
    return {
      url: `/secure-api/users?filter=${filter}&page=${page}&size=${size}&sortBy=${sortBy}&sortDirection=${sortDirection}`,
    };
  }

  toUsersPage(usersPage: UsersPageApiResponse): UsersPage {
    const res = new UsersPage();
    res.totalItems = usersPage.totalItems;
    res.firstPage = usersPage.page == 0;
    res.lastPage = usersPage.lastPage;
    res.page = usersPage.page;
    usersPage.items.forEach((user) => res.items.push(this.toUser(user)));
    return res;
  }
}
